src/sinks/http/config.rs

```diff
@@ -331,13 +332,71 @@ impl ValidatableComponent for HttpSinkConfig {
             payload_suffix: String::new(),
         };
 
-        let external_resource = ExternalResource::new(
+        let happy_external_resource = ExternalResource::new(
             ResourceDirection::Push,
-            HttpResourceConfig::from_parts(config.uri.uri.clone(), Some(config.method.into())),
-            config.encoding.clone(),
+            HttpResourceConfig::from_parts(
+                happy_config.uri.uri.clone(),
+                Some(happy_config.method.into()),
+            ),
+            happy_config.encoding.clone(),
+        );
+
+        // this config uses the Gelf serializer, which requires the "level" field to
+        // be an integer
+        let sad_config = Self {
+            uri: UriSerde::from_str("http://127.0.0.1:9000/endpoint")
+                .expect("should never fail to parse"),
+            method: HttpMethod::Post,
+            encoding: EncodingConfigWithFraming::new(
+                None,
+                GelfSerializerConfig::new().into(),
+                Transformer::default(),
+            ),
+            auth: None,
+            headers: None,
+            compression: Compression::default(),
+            batch: BatchConfig::default(),
+            request: RequestConfig::default(),
+            tls: None,
+            acknowledgements: AcknowledgementsConfig::default(),
+            payload_prefix: String::new(),
+            payload_suffix: String::new(),
+        };
+
+        let sad_external_resource = ExternalResource::new(
+            ResourceDirection::Push,
+            HttpResourceConfig::from_parts(
+                sad_config.uri.uri.clone(),
+                Some(sad_config.method.into()),
+            ),
+            sad_config.encoding.clone(),
         );
 
-        ValidationConfiguration::from_sink(Self::NAME, config, Some(external_resource))
+        // this encoder is used to ingest the input event to the input driver (which we need to succeed)
+        let sad_encoder = EncodingConfigWithFraming::new(
+            None,
+            JsonSerializerConfig::new(MetricTagValues::Full).into(),
+            Transformer::default(),
+        );
+
+        ValidationConfiguration::from_sink(
+            Self::NAME,
+            vec![
+                ComponentTestCaseConfig::from_sink(
+                    happy_config,
+                    None,
+                    Some(happy_external_resource),
+                    None,
+                ),
+                // this config only runs with the test case "encoding_error" in the yaml file.
+                ComponentTestCaseConfig::from_sink(
+                    sad_config,
+                    Some("encoding_error".to_owned()),
+                    Some(sad_external_resource),
+                    Some(sad_encoder.into()),
```

Comment:
What confuses me about this construction/pattern is that we're specifying a configuration for the component itself which uses the GELF serializer -- because we know we're going to feed it an event with a field that it considers invalid -- but then we're passing in a hand-rolled JSON serializer on the side.

Why are we doing that?

Reply:
This is described a bit on the doc comment for `ComponentTestCaseConfig::input_codec`, and in the callsite at line 449 of file runner/mod.rs (of this PR).

Essentially, we need to use an encoder for the external resources of the input driver that passes, which gets the event into the topology, so that it can then fail when under test in the actual component being tested.

Reply:
This still doesn't make sense to me.

We're testing a sink. Events basically _can't_ enter the topology in any way that would fail, because we ingest them with a `vector` source, which can handle the most arbitrarily-shaped `Event`s.

Reply:
👍 got it. the usage I have is contrived. I will look at an alternate approach to achieve what we need here.

Reply:
😓  not sure what I was thinking with that `input_codec` thing. *shakes fist at me from 8 months ago*
I removed that and just used the codec of the external resources as it was intended.